File: src/cpp/multiprocessing/multiple_processes.cpp
/**
 * Multiple Child Processes
 *
 * This example demonstrates creating and managing multiple child
 * processes from a single parent.
 *
 * Key concepts:
 * - Multiple fork() calls create multiple independent processes
 * - Each child process has its own memory and execution context
 * - Parent must wait for ALL children to prevent zombie processes
 * - Children execute concurrently on multi-core systems
 *
 * This pattern is useful for parallelizing CPU-bound work across
 * separate processes with full memory isolation.
 */

#include <chrono>
#include <iostream>
#include <sys/wait.h>
#include <thread>
#include <unistd.h>
#include <vector>

void task_a() {
    std::cout << "Task A started (PID: " << getpid() << ")\n";
    std::this_thread::sleep_for(std::chrono::seconds(1));
    std::cout << "Task A finished\n";
}

void task_b() {
    std::cout << "Task B started (PID: " << getpid() << ")\n";
    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    std::cout << "Task B finished\n";
}

int main() {
    std::cout << "Parent process (PID: " << getpid() << ")\n";

    std::vector<pid_t> children;

    // Create first child
    pid_t pid1 = fork();
    if (pid1 == 0) {
        task_a();
        return 0;
    }
    children.push_back(pid1);

    // Create second child
    pid_t pid2 = fork();
    if (pid2 == 0) {
        task_b();
        return 0;
    }
    children.push_back(pid2);

    // Parent waits for all children
    for (pid_t child : children) {
        int status;
        waitpid(child, &status, 0);
        std::cout << "Child " << child << " exited with status "
                  << WEXITSTATUS(status) << "\n";
    }

    std::cout << "All children finished\n";
    return 0;
}

File: src/python/multiprocessing/multiple_processes.py
"""
Multiple Processes Example

This script demonstrates creating and managing multiple processes that execute
different functions concurrently. Each process runs in its own memory space
with its own Python interpreter instance.

Key Concepts:
- Multiple processes run truly in parallel on multi-core systems
- Each process has isolated memory (no shared state by default)
- Process execution order is non-deterministic
- Use join() to wait for all processes to complete

Unlike threads, multiple processes bypass the GIL, enabling true parallelism
for CPU-bound tasks. However, inter-process communication requires explicit
mechanisms like Queue, Pipe, or shared memory.
"""

import os
import multiprocessing
import threading
import time


def task_foo():
    """First worker function with a 1-second execution time."""
    print("Function foo starts")
    print(f"Worker function process id: {os.getpid()}")
    print(f"Worker thread id: {threading.current_thread().ident}")
    print(f"Worker thread name: {threading.current_thread().name}")
    time.sleep(1)
    print("Function foo finishes")


def task_bar():
    """Second worker function with a 0.5-second execution time."""
    print("Function bar starts")
    print(f"Worker function process id: {os.getpid()}")
    print(f"Worker thread id: {threading.current_thread().ident}")
    print(f"Worker thread name: {threading.current_thread().name}")
    time.sleep(0.5)
    print("Function bar finishes")


def main():
    print("Main function")
    print(f"Main function process id: {os.getpid()}")
    print(f"Main thread id: {threading.current_thread().ident}")
    print(f"Main thread name: {threading.current_thread().name}")

    # Create processes for different tasks
    process_foo = multiprocessing.Process(target=task_foo)
    process_bar = multiprocessing.Process(target=task_bar)

    # Start both processes
    process_foo.start()
    process_bar.start()

    # Wait for both processes to complete
    process_foo.join()
    process_bar.join()

    print("Main function finished")


if __name__ == "__main__":
    main()
